Servers RFC3339 timestamps

ocket8888 opened the pull request:
Related: #5911 

This PR changes the `lastUpdated` field of server objects to use RFC3339 formatting to be consistent with all of the other timestamps used by that endpoint. It also makes some other tiny naming changes for consistency within the API: `cachegroup` and `cachegroupId` are now camelCase: `cacheGroup` and `cacheGroupID`. `profileNames` is now `profiles` becasue there is no other identifying information for any Profiles on server objects from which it needs to be distinguished. `physLocation` and `physLocationId` are now actual camelCase forms of "Physical Location" and "Physical Location ID", respectively. Finally, `updPending` and `revalPending` have been removed, because they provide no information that isn't trivially calculable from other properties, and are a remnant of a legacy system.

At a more implementation-detail-level, things which are never allowed to be `nil` according to API and/or database rules are now no longer allowed to be `nil` by Go's rules - because they aren't pointers anymore since they never needed to be. Also a lot of boilerplate from the method handlers was moved to a common API wrapper function.

<hr/>

## Which Traffic Control components are affected by this PR?
- Documentation
- Traffic Ops

## What is the best way to verify this PR?
Make sure the provided tests all pass.

## PR submission checklist
- [x] This PR has tests
- [x] This PR has documentation
- [x] This PR has a CHANGELOG.md entry
- [x] This PR **DOES NOT FIX A SERIOUS SECURITY VULNERABILITY**

ocket8888 commented:
TP integration tests are failing because they always fail

zrhoffman commented on traffic_ops/traffic_ops_golang/server/servers.go:
```diff
@@ -671,74 +667,47 @@ and p.id = $1
-		}
-		api.WriteRespWithSummary(w, r, v40Servers, serverCount)
-		return
+	if version.Major >= 5 {
```
This should use `Version.GreaterThanOrEqualTo`

zrhoffman commented on traffic_ops/traffic_ops_golang/server/servers.go:
```diff
@@ -1308,108 +1270,90 @@ func deleteInterfaces(id int, tx *sql.Tx) (error, error, int) {
-		if err := json.NewDecoder(r.Body).Decode(&server); err != nil {
-			api.HandleErr(w, r, tx, http.StatusBadRequest, err, nil)
-			return
+	if inf.Version.Major >= 5 {
```
This should use `Version.GreaterThanOrEqualTo`

ocket8888 commented on traffic_ops/traffic_ops_golang/server/servers.go:
```diff
@@ -1421,230 +1365,200 @@ func Update(w http.ResponseWriter, r *http.Request) {
 	where := `WHERE s.id = $1`
 	var selquery string
-	if version.Major <= 4 {
+	if inf.Version.Major <= 4 {
```
Instead of `LessThan`, `LessThanOrEqualTo`, `GreaterThan`, and `GreaterThanOrEqualTo`, we should just use `LessThan`, `GreaterThan` and logical inversions of those.

zrhoffman replied:
`LessThan` is logically the same as `!GreaterThanOrEqualTo` and `LessThanOrEqualTo` is logically the same is `!GreaterThan`, but sometimes one representation is more readable than another, and being constrained to have to use `!` any time you want `OrEqualTo` makes the code less readable

ocket8888 replied:
It's not a constraint, just a suggestion. You could just as easily use `v.GreaterThan(other) || v.Equal(other)`, or even `v.GreaterThan(other) || v == other`

zrhoffman replied:
> It's not a constraint, just a suggestion. You could just as easily use `v.GreaterThan(other) || v.Equal(other)`, or even `v.GreaterThan(other) || v == other`.

The idea that the `<=` and `>=` operators are unnecessary is very silly. Some things are just more readable using a specific operator, and if that operator doesn't exist, the code is automatically less readable as a result.

> That will be confusing only if you force yourself to attempt to read it as "v is greater than or equal to other" rather than "v is not less than other".

Any code that forces everyone to think about it in the same way as the author is not good code and should be refactored

ocket8888 replied:
You're equating `>=` and `<=` to `.GreaterThanOrEqualTo` and `LessThanOrEqualTo`, respectively. But those are literally different. These aren't operators, and they're just as equivalent to `.GreaterThan || .EqualTo` and `.LessThan || .EqualTo`. I don't care if the `OrEqualTo` versions exist, but it's crazy to me to say that, to extend your metaphor, having the pair of operators `>=` and `<` is somehow more sensible than just choosing the pair `>=` and `<=`, or `>` and `<`. There's no law that says we can't do all four, but if we're only going to have two I think it makes way more sense for them to be symmetric operations.

> Any code that forces everyone to think about it in the same way as the author is not good code and should be refactored

I couldn't agree less. Any code that leaves you with an inaccurate picture of the way the author thought about it is bad code and should be refactored. If I was stating these restrictions in English, "... an API version no less than 4.0..." seems more natural to me than "... an API version that is greater than or equal to 4.0...". If you think that's an uncommon way to phrase things, that's fair, but it's definitely because it's actually important to get the information across.

ocket8888 commented:
> You resolved [this conversation](https://github.com/apache/trafficcontrol/pull/7718#discussion_r1302074955) but went ahead and did it anyway? Please revert the Version changes not directly related to this PR. If you still think those changes are worthwhile, you can open a separate PR with those changes.

I... don't know what you mean? I resolved the conversation **because** I made changes which I believed resolved the conversation. You requested I do something, not that I not do anything?

zrhoffman commented:
> > You resolved [this conversation](https://github.com/apache/trafficcontrol/pull/7718#discussion_r1302074955) but went ahead and did it anyway? Please revert the Version changes not directly related to this PR. If you still think those changes are worthwhile, you can open a separate PR with those changes.
> 
> I... don't know what you mean? I resolved the conversation **because** I made changes which I believed resolved the conversation. You requested I do something, not that I not do anything?

From the commit message of 324fe57b9b

> Also simplified those methods to just GreaterThan, LessThan and Equal, and removed unnecessary pointer recievers and arguments (all can just be pass-by-value).

That is what I was pushing back against in https://github.com/apache/trafficcontrol/pull/7718#discussion_r1302074955 (and just now in https://github.com/apache/trafficcontrol/pull/7718#discussion_r1304603210). That part has no place in #7718, it messes with a bunch of stuff totally unrelated to using RFC 3339 timestamps for servers.

zrhoffman commented on traffic_ops/traffic_ops_golang/logs/log.go:
```diff
@@ -96,7 +96,7 @@ func Getv40(w http.ResponseWriter, r *http.Request) {
 	var result interface{}
 	var logsV5 []tc.LogV5
-	if inf.Version.GreaterThanOrEqualTo(&api.Version{
+	if !inf.Version.LessThan(api.Version{
```
Please change this back to `Version.GreaterThanOrEqualTo`

zrhoffman commented on traffic_ops/traffic_ops_golang/server/servers.go:
```diff
@@ -2281,12 +2281,12 @@ func Delete(inf *api.APIInfo) (int, error, error) {
 
 	downgraded := server.Downgrade()
-	if inf.Version.Major >= 4 {
+	if !inf.Version.LessThan(api.Version{Major: 4}) {
```
Please use `Version.GreaterThanOrEqualTo` here

zrhoffman commented on traffic_ops/traffic_ops_golang/server/servers.go:
```diff
@@ -2281,12 +2281,12 @@ func Delete(inf *api.APIInfo) (int, error, error) {
 
 	inf.CreateChangeLog(fmt.Sprintf("SERVER: %s.%s, ID: %d, ACTION: deleted", server.HostName, server.DomainName, server.ID))
-	if inf.Version.Major >= 5 {
+	if !inf.Version.LessThan(api.Version{Major: 5}) {
```
Please use `Version.GreaterThanOrEqualTo` here

zrhoffman commented on traffic_ops/traffic_ops_golang/server/servers.go:
```diff
@@ -1327,7 +1327,7 @@ func Update(inf *api.APIInfo) (int, error, error) {
 		}
 		server = tmp.Upgrade()
-	} else if inf.Version.Major >= 4 {
+	} else if !inf.Version.LessThan(api.Version{Major: 4}) {
```
Please use `Version.GreaterThanOrEqualTo` here

zrhoffman commented on traffic_ops/traffic_ops_golang/server/servers.go:
```diff
@@ -1305,7 +1305,7 @@ func Update(inf *api.APIInfo) (int, error, error) {
 	tx := inf.Tx.Tx
 
-	if inf.Version.Major >= 5 {
+	if !inf.Version.LessThan(api.Version{Major: 5}) {
```
Please use `Version.GreaterThanOrEqualTo` here

zrhoffman commented on traffic_ops/traffic_ops_golang/server/servers.go:
```diff
@@ -681,13 +681,13 @@ func Read(inf *api.APIInfo) (int, error, error) {
 		return errCode, userErr, sysErr
 	}
-	if version.Major >= 5 {
+	if !version.LessThan(api.Version{Major: 5}) {
```
Please use `Version.GreaterThanOrEqualTo` here

zrhoffman commented on traffic_ops/traffic_ops_golang/server/servers.go:
```diff
@@ -926,7 +926,7 @@ JOIN server_profile sp ON s.id = sp.server`
 			return nil, serverCount, nil, fmt.Errorf("getting servers: %w", err), http.StatusInternalServerError, nil
 		}
-		if (version.GreaterThanOrEqualTo(&api.Version{Major: 4}) && roleBasedPerms) || version.GreaterThanOrEqualTo(&api.Version{Major: 5}) {
+		if (!version.LessThan(api.Version{Major: 4}) && roleBasedPerms) || !version.LessThan(api.Version{Major: 5}) {
```
This line is particularly harder to reason about without `Version.GreaterThanOrEqualTo`, please change it back to that.

zrhoffman commented on traffic_ops/traffic_ops_golang/server/servers.go:
```diff
@@ -1477,7 +1477,7 @@ func Update(inf *api.APIInfo) (int, error, error) {
 	where := `WHERE s.id = $1`
 	var selquery string
-	if inf.Version.Major <= 4 {
+	if !inf.Version.GreaterThan(api.Version{Major: 4}) {
```
Back to the original line in question: `!Version.GreaterThan` is not as easy to reason about as `Version.LessThanOrEqualTo`. But if you feel strongly about it, `!Version.GreaterThan` it's fine, since no `Version.LessThanOrEqualTo` method already exists.

ocket8888 commented:
> > > You resolved [this conversation](https://github.com/apache/trafficcontrol/pull/7718#discussion_r1302074955) but went ahead and did it anyway? Please revert the Version changes not directly related to this PR. If you still think those changes are worthwhile, you can open a separate PR with those changes.
> > 
> > 
> > I... don't know what you mean? I resolved the conversation **because** I made changes which I believed resolved the conversation. You requested I do something, not that I not do anything?
> 
> From the commit message of [324fe57](https://github.com/apache/trafficcontrol/commit/324fe57b9b2ea01cde4db72c069fe9b0390c6cd5)
> 
> > Also simplified those methods to just GreaterThan, LessThan and Equal, and removed unnecessary pointer recievers and arguments (all can just be pass-by-value).
> 
> That is what I was pushing back against in [#7718 (comment)](https://github.com/apache/trafficcontrol/pull/7718#discussion_r1302074955) (and just now in [#7718 (comment)](https://github.com/apache/trafficcontrol/pull/7718#discussion_r1304603210)).

Well, sure, but that was after I already made the changes and resolved the conversation. So I'm confused how you could think I "went ahead and did it anyway", since when I did it there couldn't have been any objections yet to that which hadn't been done.

> That part has no place in #7718, it messes with a bunch of stuff totally unrelated to using RFC 3339 timestamps for servers.

Fair.

zrhoffman reviewed (approved):
Looks good! Presumably the TPv2 tests due to the ChromeDriver version not being the very latest.
